proj/hps_accel/gateware/hps_cfu.py

```diff
@@ -72,17 +73,39 @@ def connect_verify_register(self, m, set, get):
     def connect_macc(self, m, set, input_store, filter_store, macc, get):
         m.d.comb += macc.offset.eq(set.values[Constants.REG_INPUT_OFFSET])
 
-        m.d.comb += macc.operands.valid.eq(1)
-        m.d.comb += macc.operands.payload['inputs'].eq(Cat(
-                input_store.data_output[0].payload,
-                input_store.data_output[1].payload,
-                input_store.data_output[2].payload,
-                input_store.data_output[3].payload))
-        m.d.comb += macc.operands.payload['filters'].eq(Cat(
-                filter_store.output[0].payload,
-                filter_store.output[1].payload,
-                filter_store.output[2].payload,
-                filter_store.output[3].payload))
+        # Join all 4 input value streams and all 4 filter value streams
+        # into a single stream holding all operands.
+        # Then connect that to the macc.
+        operands_buffer = ConcatenatingBuffer([
```

Comment:
Got it. This would be cleaner. I will give it a try.

Reply:
I tried two versions of this and I don't like either of them.

Wrap a `ConcatenatingBuffer` and connect the different layouts:  
https://github.com/danc86/CFU-Playground/commit/97b64649883cf4678787d4ea04353d827da76c3d  
The wrapper class feels repetitive and ultimately pointless.

Inherit from `ConcatenatingBuffer` and re-declare the streams:
https://github.com/danc86/CFU-Playground/commit/08a449b9dfd0356d06b0910a3a582f26e395c67e
I generally avoid composition-through-inheritance, where the superclass cooperates with code defined in the subclass, because I find that style hard to read and reason about. It's why I avoided a PipelinedActor superclass the way Litex has. In this case it feels weird that the type will have both `.inputs` from the parent class but also `.input_data` and `.filter_data`. I could `del self.inputs` but that feels even weirder.

Ultimately both versions still end up with generated field names being looked up in dicts, which was your original objection. So either (a) I'm still not grokking what you have in mind, or (b) it needs some more substantial refactor of ConcatenatingBuffer that I am not seeing.

Reply:
My apologies: I really didn't intend for you to spend time addressing this comment before merging - I wanted to note it as something for us to think about and maybe come back to.

Yes, I don't love either of those option and a more substantial refactor with would be required. I understand your skepticism of inheritance - it's a widely misused mechanism. That said, the nmigen API is what it is, and there will be compromises. Perhaps that means lots of carefully constructed strings forming parts of interfaces, but I'd hope we can, over time, find ways to avoid it.